rasa/nlu/test.py

```diff
@@ -654,28 +654,19 @@ def _calculate_report(
     if report_as_dict is None:
         report_as_dict = bool(output_directory)
 
-    if output_directory:
-        report, precision, f1, accuracy = get_evaluation_metrics(
-            targets,
-            predictions,
-            output_dict=report_as_dict,
-            exclude_label=exclude_label,
-        )
+    report, precision, f1, accuracy = get_evaluation_metrics(
+        targets, predictions, output_dict=report_as_dict, exclude_label=exclude_label,
+    )
+
+    if report_as_dict:
```

Comment:
It's not clear to me why we only add confused labels if this is true.
In fact when is this actually `False`?

Reply:
This is `false` if no output directory was provided **and** it's used from the CLI. If this is `False` then the report will be formatted string (done by sklearn) which would lead to an error when trying to add the `confused_with` labels

Reply:
Is that even possible? It seems to default to a path in `run_nlu_test_async`

Reply:
True! Could still be used from the Python API though.

Reply:
Having a boolean arg which is passed so far through the code like this, changed by the value of other args, and determining the type of the value being passed back and how it's manipulated it super error prone. (Especially considering it's probably not really needed!)

Reply:
The only other option I see I is to always `report_as_dict` and purge the logging to stdout completely. In my opinion this is nothing we can do in a micro release. Also wondering if the error isn't default in the CLI and `output_directory` should be `None` by default.

Reply:
Ideally we would have always have a dictionary representation of the report at this level, and would then have different ways of outputting at the interface levels e.g. server, cli, python api. 
But agreed it's too much for now.

Reply:
True. I guess it's a bit difficult as we are currently using `sklearn` to create the formatted report instead of deciding that on our own.

Reply:
We could still always convert it to a dict when we first generate it.

Reply:
True. Should I create an issue?

Reply:
Yeah would be great 👍